### PR TITLE
Take first non-empty flux into consideration for composite reactive service discovery

### PR DESCRIPTION
**Pull request by TYsewyn**
Fixes gh-617

### CODE REVIEW

File: spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+	/**
+	 * Returns a new instance which has the additional source to be amb'd together with
+	 * the current array of sources.
+	 * <p>
+	 * This operation doesn't change the current FluxFirstEmitting instance.
+	 * @param source the new source to merge with the others
+	 * @return the new FluxFirstEmitting instance or null if the Amb runs with an Iterable
+	 */
+	@Nullable
+	FluxFirstEmitting<T> ambAdditionalSource(Publisher<? extends T> source) {
```
Comment (simonbasle) [inline spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java]: careful, you've copied that part without adapting it to the new operator. Note that since you only have a static operator, you don't need that part.

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+public class FluxFirstNonEmptyEmittingTests {
+
+	@Test(expected = NullPointerException.class)
+	public void arrayNull() {
+		CloudFlux.firstNonEmpty((Publisher<Integer>[]) null);
+	}
+
+	@Test(expected = NullPointerException.class)
+	public void iterableNull() {
+		new FluxFirstEmitting<>((Iterable<Publisher<Integer>>) null);
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: should be `new FluxFirstNonEmptyEmitting`

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	@Test
+	public void iterableOneIsNullSource() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Arrays.asList(Flux.never(),
+						(Publisher<Object>) null, Flux.never())))
+				.expectError(NullPointerException.class).verify();
+	}
+
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: this notably lacks any test that reflects what distinguish this operator from `Flux#first`: multiple tests that validate an early completing source like `Flux.empty()` vs other source(s) that emit an element. Testing the case where none of the sources emit any element would be interesting as well.

**Comment by TYsewyn** [comment]
To clarify the functionality behind `firstNotEmpty` a little bit more: we want the first flux to emit an `onNext` or `onError` signal to win. Since `Flux.first` is also taking the first completion signal into account, this is not usable for us.
One might suggest to use `Flux.first(source1.switchIfEmpty(Flux.never()), first(source2.switchIfEmpty(Flux.never()), ...)`.
Unfortunately this has a downside to it: you need to be sure that *at least* one source is able to emit an `onNext` or `onError` signal.
Since we're contacting an external source for this use case (one or more service registries), we can't guarantee this.

File: spring-cloud-commons/src/main/java/reactor/core/publisher/CloudFlux.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * in a next release.
+ *
+ * @author Tim Ysewyn
+ */
+public abstract class CloudFlux<T> extends Flux<T> {
+
+	/**
+	 * Pick the first {@link Publisher} to emit an onNext/onError signal and replay all
+	 * signals from that {@link Publisher}, effectively behaving like the fastest of these
+	 * competing sources. If none of the sources is emitting a signal, a completion signal
```
Comment (simonbasle) [inline spring-cloud-commons/src/main/java/reactor/core/publisher/CloudFlux.java]: I'd rephrase a bit: "If all the source complete empty, a single completion signal is sent. Note that if all the sources are empty (never emit an element, ie. no onNext) AND at least one is also infinite (no onComplete/onError signal), the resulting `Flux` will be infinite and empty (like `Flux#never()`)."

File: spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+					}
+
+					return true;
+				}
+			}
+			return false;
+		}
+
+		void resignFromRace() {
+			COMPETING_SUBSCRIBERS.decrementAndGet(this);
```
Comment (simonbasle) [inline spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java]: make this method return the result of the `decrementAndGet`, so that caller might directly compare it to `0` and propagate `onComplete`.

File: spring-cloud-commons/src/main/java/reactor/core/publisher/CloudFlux.java (hunk 4 above)
Comment (simonbasle) [reply spring-cloud-commons/src/main/java/reactor/core/publisher/CloudFlux.java]: to formalize the 3 cases I'm covering above:
```java
CloudFlux.firstNonEmpty(Flux.never(), Flux.empty(), Flux.just("foo")); //emits foo & completes
CloudFlux.firstNonEmpty(Flux.empty(), Flux.empty()); //completes
CloudFlux.firstNonEmpty(Flux.never(), Flux.empty()); //infinite nothing
```

File: spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+			}
+			else if (parent.tryWin(index)) {
+				won = true;
+				actual.onError(t);
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			parent.resignFromRace();
```
Comment (simonbasle) [inline spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java]: use the returned valued (see previous comment) to atomically decide to propagate complete

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Tim Ysewyn
+ */
+public class FluxFirstNonEmptyEmittingTests {
+
+	@Test(expected = NullPointerException.class)
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: NB: when/if porting to Reactor, we'll prefer the `assertThatNullPointerException` style from AssertJ be used. If you can use AssertJ in spring-cloud tests, this will be one less review item when porting 😉

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	@Test(expected = NullPointerException.class)
+	public void iterableNull() {
+		CloudFlux.firstNonEmpty((Iterable<Publisher<Integer>>) null);
+	}
+
+	@Test
+	public void firstWinner() {
+		StepVerifier
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: matter of preference but I find the `one StepVerifier method call per line` clearer to read in tests. Hopefully the code style lets you do that.

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+	@Test
+	public void bothEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.empty(), Flux.empty()))
+				.expectComplete().verifyThenAssertThat().hasNotDiscardedElements()
+				.hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void neverAndEmpty() {
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: one way to write this test and verify a longer period of time would be to use virtual time. You should also verify that no event occurs, like so:
```java
StepVerifier.withVirtualTime(CloudFlux.firstNonEmpty(Flux.never(), Flux.empty()))
    .expectSubscription()
    .expectNoEvent(Duration.ofDays(1))
    .thenCancel()
    .verifyThenAssertThat()
	.hasNotDiscardedElements().hasNotDroppedElements().hasNotDroppedErrors();
```

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+				.thenAwait(Duration.ofSeconds(2)).thenCancel().verifyThenAssertThat()
+				.hasNotDiscardedElements().hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void firstEmitsError() {
+		RuntimeException ex = new RuntimeException("forced failure");
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.<Integer>error(ex), Flux.empty()))
+				.expectError(RuntimeException.class).verifyThenAssertThat()
```
Comment (simonbasle) [inline spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: ideally you would check the message at least? can use `expectErrorMessage("forced failure")` for that.

Comment (simonbasle) [reply spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: (same in the few tests below)

File: spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+				}
+			}
+			return false;
+		}
+
+		void resignFromRace() {
+			COMPETING_SUBSCRIBERS.decrementAndGet(this);
+		}
+
+		boolean isLastRemainingFlux() {
```
Comment (simonbasle) [inline spring-cloud-commons/src/main/java/reactor/core/publisher/FluxFirstNonEmptyEmitting.java]: with the above change, this method is not necessary anymore

File: spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java (hunk 8 above)
Comment (TYsewyn) [reply spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java]: Unfortunately the code is being reformatted during checkstyle. I will take it into account when this is being ported to `reactor-core`. 😉
